### PR TITLE
Fix #1303 Create unique dirs for sorted WALs

### LINKED ISSUE #1303: Explore creating unique directories for sorted WAL recovery
When a tablet server dies, its write ahead logs need to be sorted.  A random tablet server is chosen to sort the write ahead logs.  If the random tablet server doing the sort dies or times out, the sort may be started on another write ahead log.  This may lead to multiple tablet servers sorting the write ahead logs, possibly to the same directory.  Would probably be better if each tserver doing a sort wrote to a unique dir.

### PR DESCRIPTION
**Pull request by tynyttie**
Added UniqueNameAllocator() to walPath in RecoveryPath.java.

### CODE REVIEW

**Comment by ctubbsii** [comment]
Thanks for the PR @tynyttie ; It looks like the build is failing due to formatting issues, so the tests aren't being run. Please run `mvn clean package -DskipTests` locally on your branch to format the files, and commit the changes to your branch. Then, we will be able to see the test results from in the PR checks.

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:32]: This isn't the right way to get the ServerContext. The caller should pass in the context reference that they have, as an additional parameter to this method. I see two callers for this method. `TabletServer` should have a `getContext()` method on it, and `RecoveryManager` can use `master.getContext()`.

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:52]: It would be good to log a message about the name that was chosen, before being used here, as a debug message, so that a an admin can trace the specific recovery directory back to a specific server from the server logs if necessary.

Instead of using `'/'`, you can use `'-'` so that way the directories look like `recovery-<uniquePart>`.

**Review by ctubbsii — CHANGES_REQUESTED** [review]

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:53]:
```suggestion
      log.debug("Directory selected for WAL recovery:  " + walPath);
```

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Looks good to me. I only have a minor wording suggestion for the log message.

**Review by ctubbsii — DISMISSED** [review]

**Comment by ctubbsii** [comment]
I'm going to run the full ITs on this change, but I don't anticipate any issues with this. Thanks, @tynyttie !

**Review by ctubbsii — CHANGES_REQUESTED** [review]
t seems there's a problem with recovery as a result of these changes. I didn't spend a lot of time trying to track it down, but was able to confirm that running `RestartIT` can reproduce the problem, and found the stack trace in the TServer log like:

```java
2020-12-03T00:40:54,842 [log.LogSorter] ERROR: Error during cleanup sort/copy 487f525e-2520-4706-9e15-8109f3d484af
java.io.IOException: Stream is closed!
  at org.apache.hadoop.fs.BufferedFSInputStream.getPos(BufferedFSInputStream.java:56) ~[hadoop-client-api-3.3.0.jar:?]
  at org.apache.hadoop.fs.FSDataInputStream.getPos(FSDataInputStream.java:77) ~[hadoop-client-api-3.3.0.jar:?]
  at org.apache.accumulo.tserver.log.LogSorter$LogProcessor.close(LogSorter.java:197) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
  at org.apache.accumulo.tserver.log.LogSorter$LogProcessor.sort(LogSorter.java:168) [accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
  at org.apache.accumulo.tserver.log.LogSorter$LogProcessor.process(LogSorter.java:93) [accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
  at org.apache.accumulo.server.zookeeper.DistributedWorkQueue$1.run(DistributedWorkQueue.java:113) [accumulo-server-base-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
  at org.apache.accumulo.fate.util.LoggingRunnable.run(LoggingRunnable.java:37) [accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
  at java.lang.Thread.run(Thread.java:834) [?:?]
```

It's probably just something minor where another part of our code isn't reading the correct path, but it would need to be fixed in order to merged these changes.

**Comment by tynyttie** [comment]
Looking into now.

**Comment by tynyttie** [comment]
@ctubbsii On my side, I am not seeing the test fail when I run RestartIT.

**Comment by ctubbsii** [comment]
> @ctubbsii On my side, I am not seeing the test fail when I run RestartIT.

Try this command in your branch: `mvn clean verify -Dtest=blah -Dit.test=RestartIT -PskipQA -DskipTests=false -DskipITs=false` and then watch the logs in `test/target/mini-tests/org.apache.accumulo.test.functional.RestartIT_restartMasterRecovery/logs/*` for the `ERROR` line with `LogSorter` while that is running.

**Comment by tynyttie** [comment]
Ok.Thank you.

**Comment by milleruntime** [comment]
FYI the error that @ctubbsii referenced `java.io.IOException: Stream is closed!` I think is fairly "normal" behavior for recovery.  There is usually no way to close the WAL properly if the tserver is killed so when we read the WAL and try to close it, this is the error that is seen.  This error is also seen when running RestartIT in the main branch.  There must be something else going on for the test to time out.

**Comment by ctubbsii** [comment]
Converting to draft for now, because it seems like the test failure issue is a bit difficult to track down.

**Comment by tynyttie** [comment]
Process I used for debugging this issue. Note doing this process I tailed the logs for new info. 
1. In RecoveryPath.java, I hard coded a string and notice that RestartIt passed with no issues. 
Example: walPath = Path(walPath, FileType.RECOVERY.getDirectory() + '-' + 'addedFileName');

2. In RecoveryPath.java., I used StringBuffer, StringWriter, and normal String creation to build a new string. 
Example: testSting = FileType.RECOVERY.getDirectory() + '-' 'namer.getNextName();
                Path newPath = new Path(walPath, testString); ------ RestartIT failed with this approach. 
3. In RecoveryPath.java, instead of using UniqueNameAllocator to generate the name, I create a new name generate in order to make sure the name was present when needed. ---  RestartIT failed with this approach.
4. I used Jstack to help track down the cause of threads stuck in wait state. With this approach, I couldn't pinpoint the true origin.
5. I connected to the server and added break statement to walk through the code. This approach was unsuccessful. None of my break statements was picked up by the Tserver. 
6. In LogSorter, I added a sleep where the issue is to make sure the process doesn't end. This approach was unsuccessful. 
    The issues are on: (LogSorter.java:197) and (LogSorter.java:167)

**Comment by ctubbsii** [comment]
Thanks, @tynyttie ; I spent some time looking into this, and realized that the "Stream is closed" error messages are a bit of a red herring. I've created a PR to fix those in #1856 .

Although I still haven't figured out why the changes in this PR are causing RestartIT to hang, I suspect it has something to do with the fact that it's doing recovery using a RawLocalFileSystem, which might behave differently than the default implementation of the local FileSystem.

In addition, I found a deficiency in this PR's code. It does not update the SimpleGarbageCollector code that looks for recovery files to clean up. The following patch is my attempt to fix that, but because of the hanging issue of RestartIT, I'm not sure if it will be sufficient.

<details>
<summary>Click to expand/collapse patch for SimpleGarbageCollector</summary>

```patch
diff --git a/server/base/src/main/java/org/apache/accumulo/server/ServerConstants.java b/server/base/src/main/java/org/apache/accumulo/server/ServerConstants.java
index ae6c0da8c7..1e357e5af5 100644
--- a/server/base/src/main/java/org/apache/accumulo/server/ServerConstants.java
+++ b/server/base/src/main/java/org/apache/accumulo/server/ServerConstants.java
@@ -166,8 +166,8 @@ public class ServerConstants {
     return prefix(getBaseUris(context), TABLE_DIR);
   }
 
-  public static Set<String> getRecoveryDirs(ServerContext context) {
-    return prefix(getBaseUris(context), RECOVERY_DIR);
+  public static Set<String> getRecoveryDirPatterns(ServerContext context) {
+    return prefix(getBaseUris(context), RECOVERY_DIR + "-*");
   }
 
   public static Path getInstanceIdLocation(Volume v) {
diff --git a/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java b/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java
index 40cd9d180d..da432854b0 100644
--- a/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java
+++ b/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java
@@ -393,10 +393,9 @@ public class GarbageCollectWriteAheadLogs {
   protected Map<UUID,Path> getSortedWALogs() throws IOException {
     Map<UUID,Path> result = new HashMap<>();
 
-    for (String dir : ServerConstants.getRecoveryDirs(context)) {
-      Path recoveryDir = new Path(dir);
-      if (fs.exists(recoveryDir)) {
-        for (FileStatus status : fs.listStatus(recoveryDir)) {
+    for (String dir : ServerConstants.getRecoveryDirPatterns(context)) {
+      for (FileStatus recoveryDir : fs.globStatus(new Path(dir))) {
+        for (FileStatus status : fs.listStatus(recoveryDir.getPath())) {
           try {
             UUID logId = path2uuid(status.getPath());
             result.put(logId, status.getPath());
```

</details>

**Comment by ctubbsii** [comment]
Another reason it might pass with a fixed string, but not a unique string generator, is that the unique string might need to be consistent (at least, within a process), if different parts of the recovery code are getting the path at different times.

**Comment by Manno15** [comment]
Based on what I am seeing, the scan function in `ThriftScanner` will always error with "Failed to locate tablet for table : 1 row : row_0000000000". Then it pauses and will ultimately retry. [Here is the scan function that I am seeing the breakpoints hit](https://github.com/apache/accumulo/blob/44325585c2bd8aed0a3ac87a84f44e7b6755d435/core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftScanner.java#L239)

**Comment by ctubbsii** [comment]
> Based on what I am seeing, the scan function in `ThriftScanner` will always error with "Failed to locate tablet for table : 1 row : 

The failure in the scanner (client side) and retry is expected if the tablet isn't loaded yet because it has not yet completed recovery. What needs to be determined is what is holding up the recovery on the server-side. Once the tablet is recovered and loaded on the server side, the scanner will be able to find it without issue, and read from it.

**Comment by Manno15** [comment]
That does make a lot of sense.

**Comment by Manno15** [comment]
I think I have found the issue, currently working on trying to come up with a solution. It appears everything does recover as expected (each recovery file contains the finished marker) but it doesn't properly remove it from the queue in `RecoveryManager`. This is because the destination path gets overridden with a new unique name so when it checks for that marker, it doesn't find it. At least I think that is how everything is supposed to work.

What I see now in hdfs are multiple recovery directories with the same walog UUID and all with the finished markers.

**Comment by Manno15** [comment]
I can foresee another issue possibly happening in the recover function of [TabletServer.java](https://github.com/apache/accumulo/blob/e510639958f442cd579b5e7e7fd2971b025dfae0/server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java#L1166). It also calls the same getRecoveryPath function and it will return a new directory each time. Thoughts?

**Comment by EdColeman** [comment]
Part of this reminds me of https://github.com/apache/accumulo/issues/961 - I don't think it is exactly the same, but maybe that description / solution provides pointers for things to look for? Or not....

**Comment by ctubbsii** [comment]
Another option for having each tserver use a unique directory, is to make it unique and non-random, so it is reproducible. So, if the tserver had a unique id for its lifecycle, it could just append that and it would be consistent on multiple calls to the method.

**Comment by Manno15** [comment]
I like that solution @ctubbsii . It should be pretty simple to use the name of the tserver as the unique id. 

I was also thinking of making `RecoveryPath.java` store all the recovery directories it creates so the other classes trying to get the recovery path wouldn't have too. Though if we go with your solution then that wouldn;t be necessary. I will leave that up to @tynyttie.

**Comment by milleruntime** [comment]
I think I figured out the issue with this change.  The problem is with how `RecoveryManager.recoverLogs()` works.  The first time the Master will call `RecoveryManager.recoverLogs()` it will generate `dest` (call it dest1) using `RecoveryPath.getRecoveryPath()`.  RecoveryManager will check for a "finished" file at this location, which won't exist yet and kick off the log sorting process.  The LogSorter job will complete and put a "finished" marker at dest1.  The next time the Master tries to assign the Tablet, it will call `RecoveryManager.recoverLogs()` and generate a second `dest` (call it dest2).  RecoveryManager will check for a "finished" file at dest2, won't see it and start another log sort process.  The "finished" file is never seen by the next call of `RecoveryManager.recoverLogs()` and the log sorting process will continue indefinitely.

**Comment by Manno15** [comment]
That is correct @milleruntime. Some ideas have been thrown around on how to fix it but implementing it has been more difficult since `TabletServer.java` also uses `RecoveryPath.java` for recovering tablets.

**Comment by milleruntime** [comment]
I think using the name of the tserver as the unique sub directory under the WAL directory would work.  As an added benefit, when the RecoveryManager sees that dir at the beginning of `recoveryLogs` it could also check if that tserver is still alive (coming from the master so this may be as simple as looking in a list of tservers).  It would still be possible that the tserver is alive and the thread died/hung or it was restarted so that sort process is gone.  But we could print warnings for this case and start another sort either way.

**Comment by ctubbsii** [comment]
> I think using the name of the tserver as the unique sub directory under the WAL directory would work. As an added benefit, when the RecoveryManager sees that dir at the beginning of `recoveryLogs` it could also check if that tserver is still alive (coming from the master so this may be as simple as looking in a list of tservers). It would still be possible that the tserver is alive and the thread died/hung or it was restarted so that sort process is gone. But we could print warnings for this case and start another sort either way.

The hiccup I can see with this is that I don't know what "name of the tserver" is. The most common identifier for a tserver is the hostname and port, but these are not necessarily unique (tservers fail and restart on the same host/port). I was actually just thinking about the tserver generating a unique ID internally when it starts, that can be used to make unique directories. The unique ID could be logged, or could be sent in hello messages to the manager server, or stored in Zookeeper locks, for additional uses, but for this issue, it would be enough to just log the directory that it created to be used for recovery.

Once the uniqueness issue is ironed out, with unique names, the SimpleGarbageCollector still needs to be updated and tested to ensure it works properly, cleaning up the old directories. I mocked up a patch above for that, but it has not been tested.

File: server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java
Comment (EdColeman) [inline server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java:175]: You could use parameter replacement in log statements
`log.debug("This is walPath in RecoveryManager: {}", filename);`

Comment (tynyttie) [reply server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java:175]: Thank you EdColeman, I am making the change now.

Comment (DomGarguilo) [reply server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java:175]: I see you updated this log statement in the latest commit. Looks like there may be a couple others spots that could be changed to use this as well for the sake of consistency.

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (ivakegg) [inline server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:50]: Perhaps I am missing something here, but isn't context.getInstanceID() the accumulo instance id?  How does this make the directory any more unique?

Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:50]: You're not missing anything. You're correct. This does not make it unique per tserver process.

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:53]: commons-codec, which is one of our dependencies, has a method, `DigestUtils.sha1Hex(string)`

Comment (tynyttie) [reply server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:53]: Got it. Thank you.

File: server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java:163]: Size can't be less than 0.

```suggestion
  public String getHostPort() {
    Set<TServerInstance> tserverInstances = master.onlineTabletServers();
    return tserverInstances.isEmpty() ? null : tserverInstances.stream().findFirst().get().toString();
  }

```

Comment (Manno15) [reply server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java:163]: Do you have an idea on how to properly implement this? `RecoveryManager.recoverLogs` gets called in `tabletGroupWatcher.run` when the tablets are not hosted and there are walogs. From testing, `tls.current` is null at this point. We could use `tls.last` but I am unsure if last will always be set and is prone to change.  Thoughts?

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (EdColeman) [inline server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:51]: Recently it was noted that using sha1 can be flagged by some code security scan utilities - would using sha256 or even sha512 eliminate those utilities from flagging "unsecure usages"?  This is not a security related hash so it does not matter, but those utilities don't take that into account.

The length of the string could also be shortened using something like:

```String s = new BigInteger(DigestUtils.sha256(portHost).toString(MAX_RADIX).substring(0,40)```

This would have the same number of characters as the sha1 digest, but more "bits" using base36 vs base16 - or the string length could be shortened so that it was equivalent to the 40 digit sha1 without increasing the (small) risk of a collision. This also would allow using sha512 without a ridiculous string length.

File: server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java
Comment (ctubbsii) [reply server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java]: @Manno15 If the question is directed at me: I don't, not without digging into this to come up with a solution myself.

File: server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/master/recovery/RecoveryPath.java:51]: I believe there's a variant of the sha256 method in DigestUtils that returns a hex-encoded string, that can simply be truncated directly, rather than using BigInteger, which shouldn't be necessary.

However, I'm not sure that this hashing of a location strategy is going to work at all, based on the conversation above in https://github.com/apache/accumulo/pull/1803#discussion_r568151689

**Comment by ctubbsii** [comment]
Since the original issue in #1303 was to "explore" the idea of creating unique directories, and a solution hasn't been identified in over 2 months, I'm inclined to think that this has been "explored" enough, and we should probably just close this, and the original issue #1303 , rather than continue to pursue it. ... unless somebody has some new strategy ready-to-try. Thoughts? Is #1303 worth continuing to pursue or no?

File: server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java
Comment (Manno15) [reply server/manager/src/main/java/org/apache/accumulo/master/recovery/RecoveryManager.java]: > @Manno15 If the question is directed at me: I don't, not without digging into this to come up with a solution myself. If tls.last isn't guaranteed to be the tserver that is going to be assigned this, and performing the recovery, then that would not be appropriate either.

I am not sure yet if tls.last is guranteed. I or @tynyttie  would have to do more research on a bigger cluster to get to the bottom of this. 

> it seems to me that the basic requirement is that the manager must decide the destination path, then somehow communicate that to the tserver that was assigned to do the sorting work. 

Based on what I am reading, the Tserver side of recover is only called in `Tablet.java`. This is after `RecoveryManager` is finishing do its side of te work. Not sure how everything fits in to get the full picture yet.

**Comment by Manno15** [comment]
> Since the original issue in #1303 was to "explore" the idea of creating unique directories, and a solution hasn't been identified in over 2 months, I'm inclined to think that this has been "explored" enough, and we should probably just close this, and the original issue #1303 , rather than continue to pursue it. ... unless somebody has some new strategy ready-to-try. Thoughts? Is #1303 worth continuing to pursue or no?

FWIW, I do think this is possible to implement. Either with the current mechanism, someting similar, or a rewrite of how we do the recovery process currently. Just depends on much this feature is desired and beneficial. Especially on a large scale cluster.

**Pull request closed without merging**
